lib/ecto/query/planner.ex

```diff
@@ -243,10 +243,10 @@ defmodule Ecto.Query.Planner do
     end
   end
 
-  defp prepare_source(_query, {nil, schema}, _adapter) when is_atom(schema) and schema != nil,
-    do: {schema.__schema__(:source), schema}
-  defp prepare_source(_query, {source, schema}, _adapter) when is_binary(source) and is_atom(schema),
-    do: {source, schema}
+  defp prepare_source(_query, %FromExpr{source: nil, schema: schema}, _adapter) when is_atom(schema) and schema != nil,
```

Comment:
I wonder when this scenario would happen. Wouldn't it be better to always guarantee that the source is already there when we give a schema? I think to do this we would need to change the builder/from to always include the source too.